modAionImpl/src/org/aion/zero/impl/AionHub.java

```diff
@@ -272,9 +270,23 @@ private void loadBlockchain() {
                 bestBlock != null && // recover only for non-null blocks
                 !this.repository.isValidRoot(bestBlock.getStateRoot())) {
 
+            LOG.info("Recovery initiated due to corrupt world state at block " + bestBlock.getNumber() + ".");
+
             long bestBlockNumber = bestBlock.getNumber();
             byte[] bestBlockRoot = bestBlock.getStateRoot();
 
+            // ensure that the genesis state exists before attempting recovery
+            AionGenesis genesis = cfg.getGenesis();
+            if (!this.repository.isValidRoot(genesis.getStateRoot())) {
+                LOG.info(
+                        "Corrupt world state for genesis block hash: " + genesis.getShortHash() + ", number: " + genesis
+                                .getNumber() + ".");
+
+                buildGenesis(genesis);
```

Comment:
`buildGenesis()` seems to be building the genesis state based on the current `repository`. It's correct for the latter usage, when the database is empty, but is it still valid here?

Reply:
it's valid, i.e. it always results in the same state root, because when calling `createAccount(Address)` the `stateRoot = EMPTY_TRIE_HASH;`, so it does not build on the state root of the tracked repository